### PR TITLE
[SPARK-28313][SQL] Spark sql null type incompatible with hive void type

### PR DESCRIPTION
**Pull request by ulysses-you**
## What changes were proposed in this pull request?

 [SPARK-20680](https://issues.apache.org/jira/browse/SPARK-20680), [PR](https://github.com/apache/spark/pull/17953), but actually this Jira was not solved. 

Spark is incompatible with hive void type. When table schema contains void type, spark throw exception in ddl option, like desc, show create table..

Also, spark catalog.createTable can create NullType Table that is not allowed.

Goal:

1. Support hive void column type for `desc`,`show create table` option
2. Add rule to throw exception when catalog.create NullType StructField, see [25198](https://github.com/apache/spark/pull/25198)

## How was this patch tested?

UT

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @cloud-fan @gatorsmile

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 1]
```diff
@@ -2516,4 +2516,26 @@ class HiveDDLSuite
       }
     }
   }
+
+  test("SPARK-28313: Spark sql null type incompatible with hive void type") {
+    val expectedMsg = "DataType void is not supported"
+    withTable("t") {
+      val e = intercept[AnalysisException] {
+        sql("create table t (a void)")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: how is this related to hive compatibility? It seems to me that you just change this statement from parser exception to analysis exception.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 2]
```diff
@@ -2516,4 +2516,26 @@ class HiveDDLSuite
...
+      }.getMessage
+      assert(e.contains(expectedMsg))
+    }
+
+    withTable("t") {
+      sql("CREATE TABLE t AS SELECT NULL AS col ")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: what's the behavior before your patch?

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala (hunk 1 above)
Comment (ulysses-you) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: This code is just to confirm spark cannot do this ddl that compatible with hive.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala (hunk 2 above)
Comment (ulysses-you) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: Throw a SparkException 'Cannot recognize hive type string: NULL'

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by ulysses-you** [comment]
ok to test

**Comment by ulysses-you** [comment]
ok to test

**Comment by ulysses-you** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala [hunk 3]
```diff
@@ -393,6 +393,21 @@ case class PreprocessTableInsertion(conf: SQLConf) extends Rule[LogicalPlan] {
   }
 }
 
+/**
+ * SPARK-28313: Spark sql null type incompatible with hive void type
+ */
+object CreateTableCheck extends Rule[LogicalPlan] {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: Can we make a new PR for this change? This is a behavior change, which needs more discussion and needs to add something to migration guide.

Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala]: Sure. As i say, this pr has two goals.
1. compatibility with hive
2. reject create table use NullType
So it's ok to make new pr for goal 2.

**Comment by ulysses-you** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by HyukjinKwon** [comment]
ok to test

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/NullType.scala [hunk 4]
```diff
@@ -32,6 +32,8 @@ class NullType private() extends DataType {
   override def defaultSize: Int = 1
 
   private[spark] override def asNullable: NullType = this
+
+  override def simpleString: String = "void"
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/NullType.scala]: This is also an unrelated change. I like the new naming, but it's unnecessary to the hive compatibility issue.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala [hunk 5]
```diff
@@ -961,7 +961,11 @@ private[hive] object HiveClientImpl {
   /** Get the Spark SQL native DataType from Hive's FieldSchema. */
   private def getSparkSQLDataType(hc: FieldSchema): DataType = {
     try {
-      CatalystSqlParser.parseDataType(hc.getType)
+      hc.getType match {
+        // SPARK-28313 compatible hive void type
+        case "void" => NullType
```
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: I believe this is the fix we need. But I'm curious about when can this happen if hive forbids defining void type columns.

Comment (ulysses-you) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Fix this point,  hive void type will be converted to null type. 
Such as, t1 have a void type column c1, when execute `show create table $tbl`.
Spark:
`create table t1 (c1 null)`
Hive:
`create table t1 (c1 void)`

Comment (cloud-fan) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: > But I'm curious about when can this happen if hive forbids defining void type columns.

Can hive do `create table t1 (c1 void)`?

Comment (ulysses-you) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: No, hive can not but hive can do `create table t1 as select null as c1`.

Comment (cloud-fan) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: can we make this fix surgical and only keep this change?

Comment (ulysses-you) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: How about add code to check and change null to void at `ShowCreateTableCommand.showCreateHiveTable()` ?

Comment (cloud-fan) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Does it have to be in this PR? If it has to, I'm ok with it.

Comment (ulysses-you) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/client/HiveClientImpl.scala]: Yeah~ If you have a time, you could review the latest commit.

**Comment by ulysses-you** [comment]
@cloud-fan Do you have any time to take a look ?
